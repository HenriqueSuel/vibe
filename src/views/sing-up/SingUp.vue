<template>
  <div class="register">
    <h1 class="register__title">Cadastre-se</h1>
    <KindPerson
      :isPatient="isPatient"
      v-on:change-kind-person="isPatient = $event"
    ></KindPerson>
    <section class="register__card card">
      <form class="register__form" @submit="submit">
        <input type="text" placeholder="Nome" v-model="$v.form.nome.$model" />
        <span class="mensage-error" v-if="$v.form.nome.$error"
          >Nome é obrigatorio</span
        >
        <date-picker
          v-model="$v.form.dataNascimento.$model"
          format="DD/MM/YYYY"
          placeholder="Data de Nascimento"
          :disabled-date="disabledBeforeToday"
        ></date-picker>
        <span class="mensage-error" v-if="$v.form.dataNascimento.$error"
          >Data de Nascimento é obrigatoria</span
        >
        <div class="register__form__radio">
          <span>
            <input
              id="masculino"
              type="radio"
              value="m"
              v-model="$v.form.sexo.$model"
            />
            <label for="masculino">Masculino</label>
          </span>
          <span>
            <input
              id="feminino"
              type="radio"
              value="f"
              v-model="$v.form.sexo.$model"
            />
            <label for="feminino">Feminino</label>
          </span>
        </div>

        <input
          type="text"
          placeholder="CPF"
          v-model="$v.form.doc.$model"
          v-if="isPatient"
          v-mask="'###.###.###-##'"
          :class="{ 'input-error': $v.form.doc.$error }"
        />
        <input
          type="text"
          placeholder="CRM"
          v-else
          v-model="$v.form.doc.$model"
          v-mask="'XXXXXXXXXXXXXXXXX'"
          :class="{ 'input-error': $v.form.doc.$error }"
        />
        <span class="mensage-error" v-if="$v.form.doc.$error"
          >{{ isPatient ? "CPF" : "CRM" }} Invalido</span
        >
        <button type="submit" class="button-primary">Cadastrar</button>
      </form>
    </section>
    <div class="register__login">
      <router-link to="/">Já tenho conta!</router-link>
    </div>
  </div>
</template>

<script>
import DatePicker from 'vue2-datepicker';
import KindPerson from '@/components/kind-person/KindPerson.vue';
import { required, minLength } from 'vuelidate/lib/validators';

export default {
  name: 'Login',
  components: {
    DatePicker,
    KindPerson,
  },
  data() {
    return {
      isPatient: false,
      form: {
        nome: '',
        dataNascimento: new Date(),
        sexo: 'm',
        doc: '',
      },
    };
  },
  watch: {
    isPatient() {
      this.form.doc = '';
    },
  },
  computed: {
    rules() {
      return this.isPatient ? { minLength: minLength(14) } : {};
    },
  },
  methods: {
    disabledBeforeToday(date) {
      return date > new Date();
    },
    async submit(e) {
      e.preventDefault();
      this.$v.form.$touch();
      if (this.$v.form.$error) return;
      if (this.isPatient) {
        await this.$store.dispatch('PatientStore/actionAddPatient', {
          ...this.form,
          cpf: this.form.doc,
        });
      } else {
        await this.$store.dispatch('DoctorStore/actionAddDoctor', {
          ...this.form,
          crm: this.form.doc,
        });
      }
    },
  },
  validations() {
    return {
      form: {
        nome: {
          required,
        },
        dataNascimento: {
          required,
        },
        sexo: {
          required,
        },
        doc: {
          required,
          ...this.rules,
        },
      },
    };
  },
};
</script>

<style lang="scss" scoped>
.register {
  padding: 16px;
  &__title {
    margin-top: 32px;
    text-align: center;
    color: var(--primary);
  }
  &__card {
    margin-top: 16px;
  }
  &__form {
    display: flex;
    flex-direction: column;

    &__radio {
      padding-left: 5px;
      margin-top: 16px;
      span {
        margin-right: 16px;
        input {
          margin-right: 4px;
        }
      }
    }
  }
  &__login {
    text-align: center;
    margin: 16px;
    a {
      text-decoration: none;
      color: black;
      &:hover {
        color: var(--primary);
      }
    }
  }
}
</style>
