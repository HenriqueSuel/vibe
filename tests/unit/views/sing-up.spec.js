import Vuex from 'vuex';
import { shallowMount, createLocalVue } from '@vue/test-utils';
import { VueMaskDirective } from 'v-mask';
import SingUp from '@/views/sing-up/SingUp.vue';
import Vuelidate from 'vuelidate';
import flushPromises from 'flush-promises';

const localVue = createLocalVue(Vuelidate);
localVue.use(Vuex);
localVue.use(Vuelidate);

describe('SingUp.vue', () => {
  let actions;
  let store;

  beforeEach(() => {
    actions = {
      actionAddPatient: jest.fn().mockResolvedValue({ id: '1' }),
      actionAddDoctor: jest.fn().mockResolvedValue({ id: '1' }),
    };

    store = new Vuex.Store({
      state: {},
      modules: {
        PatientStore: {
          namespaced: true,
          actions,
        },
        DoctorStore: {
          namespaced: true,
          actions,
        },
      },
    });
  });
  it('should change the input to cpf', () => {
    const wrapper = shallowMount(SingUp, {
      stubs: ['router-link'],
      directives: {
        mask: VueMaskDirective,
      },
      localVue,
      store,
    });

    wrapper.vm.$v.$touch();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('After the user submit button, $v.$invalid return true?', () => {
    const wrapper = shallowMount(SingUp, {
      stubs: ['router-link'],
      directives: {
        mask: VueMaskDirective,
      },
      localVue,
      store,
    });
    const loginBtn = wrapper.find('button');
    loginBtn.trigger('submit');
    expect(wrapper.vm.$v.$invalid).toBeTruthy();
  });

  it('should register in with the patient', async () => {
    const wrapper = shallowMount(SingUp, {
      stubs: ['router-link'],
      directives: {
        mask: VueMaskDirective,
      },
      localVue,
      store,
    });
    wrapper.setData({
      isPatient: true,
      form: {
        doc: '123.456.789-30',
        nome: 'Henrique Suel',
        dataNascimento: new Date(1996, 4, 27),
        sexo: 'm',
      },
    });
    const loginBtn = wrapper.find('button');
    loginBtn.trigger('submit');
    await flushPromises();
    expect(actions.actionAddPatient).toHaveBeenCalled();
  });

  it('should register in with the doctor', async () => {
    const wrapper = shallowMount(SingUp, {
      stubs: ['router-link', 'date-picker', 'KindPerson'],
      directives: {
        mask: VueMaskDirective,
      },
      localVue,
      store,
    });
    wrapper.setData({
      isPatient: false,
      form: {
        doc: '12345678930',
        nome: 'Henrique Suel',
        dataNascimento: new Date(1996, 4, 27),
        sexo: 'm',
      },
    });
    const loginBtn = wrapper.find('button');
    loginBtn.trigger('submit');
    await flushPromises();
    expect(actions.actionAddDoctor).toHaveBeenCalled();
  });

  it('Should return the date is valid', async () => {
    const wrapper = shallowMount(SingUp, {
      stubs: [],
      localVue,
      store,
    });

    const isDateValid = wrapper.vm.disabledBeforeToday(new Date(1996, 6, 27));
    expect(isDateValid).toBeFalsy();
  });

  it('Should return that data is invalid', async () => {
    const wrapper = shallowMount(SingUp, {
      stubs: [],
      localVue,
      store,
    });

    const isDateValid = wrapper.vm.disabledBeforeToday(new Date(2030, 6, 27));
    expect(isDateValid).toBeTruthy();
  });
});
