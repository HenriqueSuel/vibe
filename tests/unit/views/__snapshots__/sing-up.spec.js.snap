// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SingUp.vue should change the input to cpf 1`] = `
<div class="register">
  <h1 class="register__title">Cadastre-se</h1>
  <kindperson-stub></kindperson-stub>
  <section class="register__card card">
    <form class="register__form"><input type="text" placeholder="Nome">
      <!---->
      <date-picker-stub value="Sun Sep 27 2020 23:43:58 GMT-0300 (GMT-03:00)" defaultvalue="Sun Sep 27 2020 00:00:00 GMT-0300 (GMT-03:00)" disableddate="function () { [native code] }" type="date" getclasses="function _default() {
          return [];
        }" titleformat="YYYY-MM-DD" format="DD/MM/YYYY" timetitleformat="YYYY-MM-DD" disabledtime="function _default() {
          return false;
        }" hourstep="1" minutestep="1" secondstep="1" scrollduration="100" valuetype="date" rangeseparator=" ~ " placeholder="Data de Nascimento" editable="true" clearable="true" prefixclass="mx" inputclass="mx-input" inputattr="[object Object]" appendtobody="true" popupstyle="[object Object]" confirmtext="OK" shortcuts=""></date-picker-stub>
      <!---->
      <div class="register__form__radio"><span><input id="masculino" type="radio" value="m"> <label for="masculino">Masculino</label></span> <span><input id="feminino" type="radio" value="f"> <label for="feminino">Feminino</label></span></div> <input type="text" placeholder="CRM" class="">
      <!----> <button type="submit" class="button-primary">Cadastrar</button>
    </form>
  </section>
  <div class="register__login">
    <router-link-stub to="/">JÃ¡ tenho conta!</router-link-stub>
  </div>
</div>
`;
